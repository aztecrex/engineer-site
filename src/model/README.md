# Model


## Responsibilities

### Store

The store is exported as `store`.  It initially contains the default
state of the application. The reducer is modular and delegated
to submodules such as `articles.js`. Each submodule provides a reducer
that produces default state when invoked with no arguments. Submodule
reducers are combined with `Redux.combineReducers` to create the
final reducer to initialize the store.

### Action Constructors

The action constructors are exported as `acions`. Actions are an object
with each key mapped to an action constructor.

The action constructors are gathered from submodules. By convention,
submodules follow the same convention as the main module and expose
their actions through an `actions` export. All sub-module actions are
merged by the module so it is important that each submodule use
unique constructor names.

## Testing

Testing is performed by each submodule on its own reducer and actions.
We currently do not test the composition since making it testable could
reduce readability and it is pretty straightforward wiring.

In the future, I might figure out how to test the wiring and maintain
readability.
