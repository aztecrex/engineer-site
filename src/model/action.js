
const flipName = () => ({
  type: 'FLIP_NAME'
});

export { flipName };
