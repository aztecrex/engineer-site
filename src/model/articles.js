import R from 'ramda';

const receiveContentType = 'ARTICLES_RECEIVE_CONTENT';
const needContentType = 'ARTICLES_NEED_CONTENT';


const latestFirst =
  R.comparator((entry1,entry2) => entry1.published > entry2.published);

const parsePublished = elem => R.assoc('published', new Date(elem.published), elem);

const prepareEntries = entries =>
  R.map(parsePublished, entries).sort(latestFirst);

const entryToIndex = entry => ({[entry.id]: entry});
const prepareIndex = R.compose(R.mergeAll, R.map(entryToIndex));

const defaultState = entries => ({
  directory: prepareEntries(entries),
  index: prepareIndex(entries),
  content: {}
});

const createReducer = entries => (state = defaultState(entries), action) => {
  let newState;
  if (action) {
    switch(action.type) {
      case receiveContentType:
        let lens = R.lensPath(['content',action.digest]);
        newState = R.set(lens, action.content, state);
        break;
      default:
        newState = state;
        break;
    }
  } else newState = state;
  return newState;
};

const receiveContent = (digest, content) => {
  return {
    type: receiveContentType,
    digest,
    content
  };
};

const needContent = (digest) => {
  return {
    type: needContentType,
    digest
  };
};

const actions = {
  receiveContent,
  needContent
};

export {
  createReducer,
  actions
};
