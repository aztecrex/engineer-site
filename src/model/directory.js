import R from 'ramda';

const newestFirst = (entry1,entry2) => {
  let d1 = new Date(entry1.published);
  let d2 = new Date(entry2.published);
  let comp = d1 > d2 ? -1 : (d2 > d1 ? 1 : 0);
  return comp;
};

const parsePublished = elem => R.assoc('published', new Date(elem.published), elem);

const prep = entries =>
  R.map(parsePublished, entries.concat().sort(newestFirst));

export default prep;
