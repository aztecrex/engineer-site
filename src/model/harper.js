const flipNameType = 'HARPER_FLIP_NAME';

const flipName = () => ({
  type: flipNameType
});

const reduce = (state = "Harper", action) => {

  let newState;
  if (action) {
    switch (action.type) {
      case flipNameType:
        newState = state.split('').reverse().join('');
        break;
      default:
        newState = state;
        break;
    }
  } else newState = state;

  return newState;
};

const actions = {flipName};

export { reduce, actions };
