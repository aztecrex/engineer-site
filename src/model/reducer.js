import entries from '../articles/directory.json';
import makeArticles from './articles';

const defaultState = {
  name: "Harper",
  directory: makeArticles(entries).reduce().directory // not the correct way yet
};

const reduce = (state = defaultState, action) => {

  let newState;
  switch (action.type) {
    case 'FLIP_NAME':
      newState = {
        ...state,
        name: state.name.split('').reverse().join('')
      };
      break;
    default:
      newState = state;
      break;
  }
  return newState;
};

export { reduce };
