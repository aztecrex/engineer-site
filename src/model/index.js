import { createStore } from 'redux';

import { reduce } from './reducer';
import { flipName } from './action';

const store = createStore(reduce);

const actions = { flipName };

export { actions, store };
