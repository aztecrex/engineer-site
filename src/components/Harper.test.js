import React from 'react';
import ezJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
const createMockStore = configureMockStore([]);

import {actions} from '../model';
const {flipName} = actions;

import ConnectedHarper, {Harper, mapStateToProps} from './Harper';

describe('unconnected', () => {
  it('renders happy', () => {

    // when
    const rendered = shallow(<Harper name='George' />);

    // then
    expect(ezJson(rendered)).toMatchSnapshot();

  });

  it ('renders missing', () => {

    // when
    const rendered = shallow(<Harper />);

    // then
    expect(ezJson(rendered)).toMatchSnapshot();

  });
});

describe('map state', () => {

  it('extracts name only', () => {
    // given
    const state = {name:'Nancy', extra:'stuff'};

    // when
    const actual = mapStateToProps(state);

    // then
    expect(actual).toEqual({name:'Nancy'});

  });

  it('is ok with missing prop', () => {
    // given
    const state = {extra:'stuff'};

    // when
    const actual = mapStateToProps(state);

    // then
    expect(actual).toEqual({name:''});

  });

  it('is ok with missing state', () => {

    // when
    const actual = mapStateToProps();

    // then
    expect(actual).toEqual({name:''});

  });
});

describe('connected', () => {

  it('renders mapped state', () => {

    // given
    const state = {name:'Lena'};
    const store = createMockStore(state);

    // when
    const rendered = shallow(<ConnectedHarper store={store} />);

    // then
    expect(ezJson(rendered)).toMatchSnapshot();


  });

  it('issues action on click', () => {

    // given
    const store = createMockStore({});
    const rendered = mount(<ConnectedHarper store={store} />);

    // when
    rendered.simulate('click');

    // then
    expect(store.getActions().length).toEqual(1);
    expect(store.getActions()[0]).toEqual(flipName());

  });

});
