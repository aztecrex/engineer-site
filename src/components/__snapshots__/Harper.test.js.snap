exports[`unconnected renders happy 1`] = `
<span
  className="harper">
  George
</span>
`;

exports[`unconnected renders missing 1`] = `
<span
  className="harper" />
`;
